File: 0225_Implement_Stack_using_Queues/solution.py
'''
225. Implement Stack using Queues

Implement the following operations of a stack using queues.

push(x) -- Push element x onto stack.
pop() -- Removes the element on top of the stack.
top() -- Get the top element.
empty() -- Return whether the stack is empty.
Notes:
You must use only standard operations of a queue -- which means only push to back, peek/pop from front, size, and is empty operations are valid.
Depending on your language, queue may not be supported natively. You may simulate a queue by using a list or deque (double-ended queue), as long as you use only standard operations of a queue.
You may assume that all operations are valid (for example, no pop or top operations will be called on an empty stack).
'''


class Stack(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.queue = []

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        self.queue.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        for x in range(len(self.queue) - 1):
            self.queue.append(self.queue.pop(0))
        return self.queue.pop(0)

    def top(self):
        """
        :rtype: int
        """
        top = None
        for x in range(len(self.queue)):
            top = self.queue.pop(0)
            self.queue.append(top)
        return top

    def empty(self):
        """
        :rtype: bool
        """
        return len(self.queue) <= 0

File: 0225_Implement_Stack_using_Queues/solution.cpp
/*
225. Implement Stack using Queues

Level: Easy

https://leetcode.com/problems/implement-stack-using-queues

- 232. Implement Queue using Stacks
*/
#include <vector>
#include <string>
#include <queue>
#include <unordered_set>

#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class MyStack
{
private:
  queue<int> qData;
  queue<int> qHelper;

public:
  /** Initialize your data structure here. */
  MyStack()
  {
  }

  /** Push element x onto stack. */
  void push(int x)
  {
    qData.push(x);
  }

  /** Removes the element on top of the stack and returns that element. */
  int pop()
  {
    while (qData.size() > 1)
    {
      qHelper.push(qData.front());
      qData.pop();
    }
    int element = qData.front();
    qData.pop();
    while (!qHelper.empty())
    {
      qData.push(qHelper.front());
      qHelper.pop();
    }
    return element;
  }

  /** Get the top element. */
  int top()
  {
    while (qData.size() > 1)
    {
      qHelper.push(qData.front());
      qData.pop();
    }
    int element = qData.front();
    qHelper.push(qData.front());
    qData.pop();
    while (!qHelper.empty())
    {
      qData.push(qHelper.front());
      qHelper.pop();
    }
    return element;
  }

  /** Returns whether the stack is empty. */
  bool empty()
  {
    return qData.empty();
  }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */